src/JsonEncodeException.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Crell\ApiProblem;
+
+use Throwable;
+
+class JsonEncodeException extends AbstractJsonException
+{
+    /**
+     * @var mixed
+     */
+    protected $json;
+
+    public function __construct(string $message = '', int $code = 0, Throwable $previous = null, $json = null)
+    {
+        parent::__construct($message, $code, $previous);
+        $this->setJson($json);
+    }
+
+    public function setJson($json) : self
+    {
+        $this->json = $json;
+        return $this;
+    }
+
+    public function getJson()
+    {
+        return $this->json;
+    }
+
+    public static function fromJsonError(int $jsonError, $json): self
+    {
+        return new self(static::getExceptionMessage($jsonError), $jsonError, null, $json);
+    }
```

Comment:
The problem is strict types. `$json` parameter is different for each class so they are different.

So we could:
- leave it like that
- use `$json` as mixed type (in both classes)
- use just `fromJsonError(int $jsonError)` in the parent class and add an immutable method like `withJson()` for each different class, but `getJson()` should return a nullable value.

What do you think?

Reply:
Hrm.  OK, I see what you mean.  In that case... $json in `JsonEncodeException` isn't actually a JSON string; it would be the array/string that failed to be turned into JSON.  So it should be named something else (as should its set/get methods), like `source` or something.